wit/types.wit

```diff
@@ -16,29 +25,29 @@ interface messaging-types {
     /// The interface doesn't highlight this difference in the type itself as that's uniquely a consumer issue.
     type channel = string;
   
-    /// Configuration includes a required list of channels the guest is subscribing to, and an optional list of extensions key-value pairs 
-    /// (e.g., partitions/offsets to read from in Kafka/EventHubs, QoS etc.).
+    /// Configuration includes a required list of channels the guest is subscribing to, and an
+    /// optional list of extensions key-value pairs (e.g., partitions/offsets to read from in 
+    /// Kafka/EventHubs, QoS etc.).
     record guest-configuration {
         channels: list<channel>,
         extensions: option<list<tuple<string, string>>>
     }
   
-    /// Format specification for messages 
-    ///  - more info: https://github.com/clemensv/spec/blob/registry-extensions/registry/spec.md#message-formats
-    ///  - message metadata can further decorate w/ things like format version, and so on.
-    enum format-spec {
-        cloudevents,
-        http,
-        amqp,
-        mqtt,
-        kafka,
-        raw
-    }
-  
-    /// A message with a binary payload, a format specification, and decorative metadata.
+    /// A message with a binary payload and additional information
     record message {
+        /// The topic/subject/channel this message was received or should be sent on
+        topic: channel,
+        /// An optional content-type describing the format of the data in the message. This is 
+        /// sometimes described as the "format" type
+        content-type: option<string>,
+        /// An optional topic for use in request/response scenarios. Senders and consumers of
+        /// messages must not assume that this field is set and should handle it in their code
+        /// accordingly.
+        reply-to: option<string>,
+        /// An opaque blob of data
```

Comment:
Is the intention to allow the recipient of the message to `reply-to` someone other than the sender of the message?  If "yes", then the blocking `request-reply.request` function seems like a footgun, since it blocks until it gets a reply (which might not come back to it).  If "no", then do we even need a field or can we just say that the `request-reply.request` function automatically sets it on the outgoing side, and then perhaps we could have a `reply` method on the `message` resource (as suggested above) which can be trusted to reply to the correct sender.

Reply:
I actually like that idea quite a bit. I was wondering if implementations of the request function should just automatically set the reply-to field, but having it be an implementation detail of `request` is probably a good idea.

I probably wouldn't have a `reply` method on the resource because we still want it to be optional, so is there a way to have a `reply` method in `request-reply` that takes a `borrow<message>` in order to send a response?

Reply:
Yes, a `reply` method in `request-reply` could work.  So, iiuc, are you thinking an imported interface like:
```wit
interface request-reply {
  reply: func(reply-to: borrow<message>, reply: message) -> result<_, error>;
}
```
(and I suppose if you call `reply` on a `message` that was not in fact sent via `request` it returns an `error`)?

Thinking about this some more, I have another high-level goals question for you: is it a goal for a runtime to be able to opportunistically link two components together when the sending and receiving component are on the same machine?  This would be like [what we're trying to do in 0.3 with WASI HTTP](https://docs.google.com/presentation/d/1W62iGDiFLSBw6gZU1y8Ec7j0b_5MXOB4Yq448bRr_XQ), but with `wasi:messaging/types.message` in place of `wasi:http/types.{request,response}`.  WASI HTTP has to wait until 0.3 because it's all async/streamy, but with the totally synchronous request-reply design proposed here, we could perhaps get away with it in 0.2.  However, I find that that this goal of "direct component linking" ends up having interesting design implications (which we've incorporated into WASI HTTP) since now you don't have any host code sitting in-between two components (e.g., you can't rely on the host to set the `reply-to` field, since the host isn't sitting between the two components).

Reply:
Yeah that function was exactly what I was thinking. I think direct composition could work as well if the `reply-to` field no longer appears in the message and is handled by the implementation of the resource

Reply:
The tricky thing I'm having trouble seeing with direct-composition is: if we want direct composition (without magic middleware) we need to have 1 interface that can be both imported and exported (analogous to `wasi:http/handler`; say: `wasi:messaging/handler`).  Once we do this, the natural function signature for the caller (which returns a `message`) must also match that of the callee, which means the callee needs to reply by returning a `message`, not calling an imported `reply` function.  But now this suggests that we need *two* different exportable interfaces: one for request-reply flows (where there is a `message` return value) and one for reply-less messages (with just a `status` return value).  Maybe this is a good thing: I expect handling a request/reply message ends up feeling rather different to a developer than handling a reply-less message and so having different entry points perhaps makes this distinction more clearly in the code (and avoids certain classes of runtime errors).  WDYT?

Reply:
Yeah I think they are handled in different ways and so having the two different interfaces seems to be the way to go if I understood what you said properly